### DISCUSSION TITLE
How can I update my column to make it unique ?

### DISCUSSION THREAD

**Discussion by Hakamate** [adonisjs/core#1396] [Help] [answered]
Hello everyone,

So I made a migration like that : 
```
  protected tableName = 'users'

  public async up () {
    this.schema.createTable(this.tableName, (table) => {
      table.increments('id').primary()
      table.string('email', 255).notNullable()
      table.string('password', 180).notNullable()
      table.string('remember_me_token').nullable()
      table.timestamps(true)
    })
  }
```

But I want to make "email" column unique : 

`make:migration add_unique_to_email --table=users`

That generate : 
```
import BaseSchema from '@ioc:Adonis/Lucid/Schema'

export default class Users extends BaseSchema {
  protected tableName = 'users'

  public async up () {
    this.schema.table(this.tableName, (table) => {
      table.string('email', 255).unique().notNullable()
    })
  }

  public async down () {
    this.schema.table(this.tableName, (table) => {
      table.string('email', 255).notNullable()
    })
  }
```

and running migration print this : 
`×  fatal       error: alter table "users" add column "email" varchar(255) not null - « email » colum from « users » relation already exist`

I don't want to add but update column

**Comment by iammrsea** [comment]
Try this:
```javascript
table.string('email').unique().alter()
or 
table.string('email').alter().unique()
```
I am not it's going to work though

**Reply by Hakamate** [reply]
Thank but that not working :
`×  fatal       error: alter table "users" add constraint "users_email_unique" unique ("email") - could not create unique index "users_email_unique"`

Do i have to do something in my database or Model ?

**Reply by iammrsea** [reply]
Perhaps you can just drop the column and add it again. Something like this:
```javascript
table.dropColumn('email');
table.string('email').unique().notNullable()
```

**Reply by Hakamate** [reply]
Yes that will work, but if I drop the column, all data will be deleted 😅

**Reply by iammrsea** [reply]
Oh my bad, didn't know you already had some important data in your database. Perhaps, you can ensure the uniqueness of the column on application level using  validator. Something like this:
```javascript
await request.validate({
schema: schema.create({
email: schema.string({trim:true},[
rules.email(),
rules.unique({table:'users',column:'email})
]),
...
})
})
```
With that, the email column will remain unique

**Reply by Hakamate** [reply]
That works but is just a validation before register. There is any way to just put a unique constraint by updating colomn ?

(Thanks for all your responses, all works :D)

**Comment by kennedyrs** [comment, marked as answer]
AdonisJS uses Knex behind, take a look at [Knex documentation](http://knexjs.org/#Schema-unique) you can use this method.

If you have any email record duplicated at the database in the moment of the alter, it will not work

**Reply by Hakamate** [reply]
I saw Knex, but yes i will use it if i can't find other solutions.

Thanks !
